src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+        private BufferSegment _readHead;
+        private int _readHeadIndex;
+
+        // The commit head which is the extent of the bytes available to the IPipelineReader to consume
+        private BufferSegment _commitHead;
+        private int _commitHeadIndex;
+
+        // The write head which is the extent of the IPipelineWriter's written bytes
+        private BufferSegment _writingHead;
+
+        private PipeReaderState _readingState;
+
+        private bool _disposed;
+
+        private DefaultPipeReader _reader;
+        private DefaultPipeWriter _writer;
+
+        internal long Length => _length;
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> using <see cref="PipeOptions.Default"/> as options.
+        /// </summary>
+        public Pipe() : this(PipeOptions.Default)
+        {
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> with the specified <see cref="PipeOptions"/>.
+        /// </summary>
+        public Pipe(PipeOptions options)
+        {
+            if (options == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.options);
+            }
+
+            _bufferSegmentPool = new BufferSegment[SegmentPoolSize];
+
+            _readingState = default;
+            _readerCompletion = default;
+            _writerCompletion = default;
+
+            _pool = options.Pool;
+            _minimumSegmentSize = options.MinimumSegmentSize;
+            _maximumSizeHigh = options.PauseWriterThreshold;
+            _maximumSizeLow = options.ResumeWriterThreshold;
+            _readerScheduler = options.ReaderScheduler ?? PipeScheduler.Inline;
+            _writerScheduler = options.WriterScheduler ?? PipeScheduler.Inline;
+            _readerAwaitable = new PipeAwaitable(completed: false);
+            _writerAwaitable = new PipeAwaitable(completed: true);
+            _reader = new DefaultPipeReader(this);
+            _writer = new DefaultPipeWriter(this);
+        }
+
+        private void ResetState()
+        {
+            _readerCompletion.Reset();
+            _writerCompletion.Reset();
+            _commitHeadIndex = 0;
+            _currentWriteLength = 0;
+            _length = 0;
+        }
+
+        /// <summary>
+        /// Allocates memory from the pipeline to write into.
+        /// </summary>
+        /// <param name="minimumSize">The minimum size buffer to allocate</param>
+        internal Memory<byte> GetMemory(int minimumSize)
+        {
+            if (_writerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (minimumSize < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumSize);
+            }
+
+            lock (_sync)
+            {
+                BufferSegment segment = _writingHead ?? AllocateWriteHeadUnsynchronized(minimumSize);
```

Comment:
Could this be optimized if the default scheduler is used? ... in which case the reader is reading on the same thread.

Reply:
Even if scheduler is inline you don't have any guarantees what thread would call into pipe

Reply:
Yes, but at that point you could do just a simple interlocked read/write, correct? Anyway, I think the code is good as is, I just noticed how much heavy locking we do and it looks like we could optimize this code in the future.

Reply:
It's actually really hard to remove the locking, we tried several time and failed, each time we ended up locking more 😄 . That said, I'm happy to look at this again once we have lots of tests in place.

Reply:
OK. Let's see if we can optimized the locks later.

Reply:
Is it a valid use case for Pipe to have multiple consumers or producers anyway? Seems like that leads to a lot of undefined behavior down the road.

Reply:
No, pipe is single producer single consumer.